tests/unit/s2n_optional_client_auth_test.c

```diff
@@ -0,0 +1,465 @@
+    s2n_blocked_status server_blocked;
+    do {
+        int rc;
+        rc = s2n_negotiate(client_conn, &client_blocked);
+        if (rc != 0 && (client_blocked && errno != EAGAIN)) {
+            return -1;
+        }
+        rc = s2n_negotiate(server_conn, &server_blocked);
+        if (rc != 0 && (server_blocked && errno != EAGAIN)) {
+            return -1;
+        }
+
+        tries += 1;
+        if (tries >= MAX_TRIES) {
+            return -1;
+        }
+    } while (client_blocked || server_blocked);
+    return 0;
+}
+
+int main(int argc, char **argv)
+{
+    struct s2n_config *client_config;
+    struct s2n_config *server_config;
+    const struct s2n_cipher_preferences *default_cipher_preferences;
+    char *cert_chain_pem;
+    char *private_key_pem;
+    char *dhparams_pem;
+
+    BEGIN_TEST();
+
+    /* s2n support for mutual auth when in FIPS mode is not yet implemented. */
+    if (s2n_is_in_fips_mode()) {
+        /* Test optional client auth fails using s2n_config_set_client_auth_type. */
+        EXPECT_NOT_NULL(server_config = s2n_config_new());
+        EXPECT_FAILURE(s2n_config_set_client_auth_type(server_config, S2N_CERT_AUTH_OPTIONAL));
+        EXPECT_SUCCESS(s2n_config_free(server_config));
+
+        /* Test optional client auth fails using s2n_connection_set_client_auth_type. */
+        struct s2n_connection *conn;
+        EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_SERVER));
+        EXPECT_FAILURE(s2n_connection_set_client_auth_type(conn, S2N_CERT_AUTH_OPTIONAL));
+        EXPECT_SUCCESS(s2n_connection_free(conn));
+
+        END_TEST();
+    }
+
+    EXPECT_SUCCESS(setenv("S2N_ENABLE_CLIENT_MODE", "1", 0));
+    EXPECT_NOT_NULL(cert_chain_pem = malloc(S2N_MAX_TEST_PEM_SIZE));
+    EXPECT_NOT_NULL(private_key_pem = malloc(S2N_MAX_TEST_PEM_SIZE));
+    EXPECT_NOT_NULL(dhparams_pem = malloc(S2N_MAX_TEST_PEM_SIZE));
+
+    /* Setup baseline server config and certs. */
+    EXPECT_NOT_NULL(server_config = s2n_config_new());
+    EXPECT_SUCCESS(s2n_read_test_pem(S2N_DEFAULT_TEST_CERT_CHAIN, cert_chain_pem, S2N_MAX_TEST_PEM_SIZE));
+    EXPECT_SUCCESS(s2n_read_test_pem(S2N_DEFAULT_TEST_PRIVATE_KEY, private_key_pem, S2N_MAX_TEST_PEM_SIZE));
+    EXPECT_SUCCESS(s2n_read_test_pem(S2N_DEFAULT_TEST_DHPARAMS, dhparams_pem, S2N_MAX_TEST_PEM_SIZE));
+    EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key(server_config, cert_chain_pem, private_key_pem));
+    EXPECT_SUCCESS(s2n_config_add_dhparams(server_config, dhparams_pem));
+    EXPECT_NOT_NULL(default_cipher_preferences = server_config->cipher_preferences);
+
+
+    /*
+     * Test optional client auth using **s2n_config_set_client_auth_type** with a valid client cert provided.
+     */
+
+    EXPECT_NOT_NULL(client_config = s2n_config_new());
+    EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key(client_config, cert_chain_pem, private_key_pem));
+    EXPECT_SUCCESS(s2n_config_set_verify_cert_chain_cb(client_config, &accept_all_rsa_certs, NULL));
```

Comment:
Side note, unrelated to your change: We're going to be removing the `s2n_config_set_verify_cert_chain_cb` API call soon once https://github.com/awslabs/s2n/pull/618 gets merged in (probably within the next few days), and will instead use a `verify_hostname_fn` callback.

Depending on which change gets merged in first, one of these PR's will need to fix the merge conflicts.